### ISSUE TITLE
S3 Bucket Upload BuildStep

### ISSUE THREAD

**Issue by kallisti5** [buildbot/buildbot#4312]
It would be *really* nice to have a native S3 bucket upload within buildbot.

Since the directly HTTP upload to the buildmaster doesn't work well for large artifacts, having steps to upload one or more files to S3 directly from a worker would be a *great* workaround for a lot of people.

Ideally the step would consume:

  * s3 server (or default amazon server)
  * s3 region (or none)
  * s3 key  (via Secret!)
  * s3 secret (via Secret!)
  * s3 bucket name
  * source file
  * destination file path

boto3 is a common Python module for s3 support.

**Comment by kallisti5** [comment]
A good (and cheap!) service to test s3 is wasabi.com (or you can deploy minio.io locally)

**Comment by tardyp** [comment]
I agree this would be good to have an S3Upload step, even better, within an artifact upload framework. Like the secretManager, you can make generic the file globing stuff, that already exists in Upload steps.

Difficulty is that this needs to be in the worker code. We try to avoid putting smarts in the worker, because it is much harder to upgrade.
Another option is to push some binaries to the worker and make use of it (an uploader made with pyinstaller, or golang https://github.com/gerlacdt/aws-s3-down-uploader). THis option is implementable without needing new function in worker code.

**Comment by kallisti5** [comment]
This is actually our workaround for now (using a local s3 client hand installed and configured with secrets on each builder by each builder owner)

However, it seems like a really common use case that most build systems implement... Jenkins, Bamboo, etc.
